src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            new string[] { typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName }.ToImmutableArray();
+
+        protected abstract SyntaxNode GetForeachStatement(SyntaxToken token);
+        protected abstract (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode foreachStatement);
+        protected abstract void ConvertToForStatement(SemanticModel model, ForeachInfo info, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var foreachStatement = GetForeachStatement(token);
+            if (foreachStatement == null)
+            {
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var foreachInfo = GetForeachInfo(semanticFact, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForeachToForCodeAction(
+                    FeaturesResources.Convert_foreach_to_for,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected string CreateUniqueMethodName(SemanticModel model, SyntaxNode contextNode, string baseName)
+        {
+            Contract.ThrowIfNull(contextNode);
+            Contract.ThrowIfNull(baseName);
+
+            return NameGenerator.GenerateUniqueName(baseName, string.Empty,
+               name => model.LookupSymbols(contextNode.SpanStart, /*container*/null, name).Length == 0);
+        }
+
+        protected T AddElasticAnnotation<T>(T node, SyntaxTrivia elasticTrivia) where T : SyntaxNode
+        {
+            var first = node.GetFirstToken();
+            node = node.ReplaceToken(first, first.WithPrependedLeadingTrivia(elasticTrivia));
+
+            var last = node.GetLastToken();
+            return node.ReplaceToken(last, last.WithAppendedTrailingTrivia(elasticTrivia));
+        }
+
+        protected T AddRenameAnnotation<T>(T node, string name) where T : SyntaxNode
+        {
+            var token = node.DescendantTokens().FirstOrDefault(t => t.Text == name);
+            if (token == default(SyntaxToken))
+            {
+                return node;
+            }
+
+            return node.ReplaceToken(token, token.WithAdditionalAnnotations(RenameAnnotation.Create()));
```

Comment:
this whole approach feels very wrong.  We generated the tree manually.  Why did we not just put the rename annotation on the name when we generated things?

Reply:
because of of nodes I parse rather than do time consuming SyntaxFactory build up.

Reply:
Yes.  This has now become *more* complicated because you have sidestepped the standard APIs all the rest of our code-actions use.  Please follow the patterns and practices of the majority of the IDE codebase.  We want it to stay consistent and uniform.